iree/samples/emitc_modules/CMakeLists.txt

```diff
@@ -17,13 +17,24 @@ if(${IREE_ENABLE_EMITC})
       "-iree-hal-target-backends=vmvx"
   )
 
+  iree_cc_library(
```

Comment:
We should be able to avoid the need for two rules per module.
You can use COPTS instead of defines to add the impl flag and they'll only be used for the .c files in the library.

To keep the compiler simple with 1 file output we can have an `iree/vm/module_impl_emitc.c` that contained:
```c
#include EMITC_IMPLEMENTATION
```
and then add
```cmake
HDRS
  "add_module.h"
SRCS
  "iree/vm/module_impl_emitc.c"
COPTS
  -DEMITC_IMPLEMENTATION=add_module.h
```

If the header guard in the generated files is changed to `#if defined(EMITC_IMPLEMENTATION)` then this should be all that's needed. Someone wanting to have multiple implementations in the same file could do so by writing it themselves:
```c
#define EMITC_IMPLEMENTATION
#include "add_module.h"
#include "sub_module.h"
#include "div_module.h"
```

Reply:
(the nice thing about this too is that if not in iree_cc_library land a user can easily include all the modules without any need for build system goo)

Reply:
(from discord - sorry for being hand-wavy!)

I think the iree_c_module rule can effectively be an alias for iree_cc_library, and other iree_cc_libraries should be able to take a dep on them directly. If iree_c_module generates the header file and then just exposes it as above with HDRS/SRCS/COPTS on that iree_cc_library it should all just work in the IREE world.

Outside of IREE, a user can generate the .h however they want and manually include the files (in their main .c file, etc) and don't need any of this stuff, but when in the IREE world it's much nicer to have all the generation/deps/etc nicely hidden.

Reply:
More concretely, it can look like https://github.com/google/iree/blob/552d3f888e46f96c8227b1bd2e4600b85338237b/build_tools/cmake/iree_c_embed_data.cmake#L81-L87

Reply:
Really nice proposal to solve this @benvanik! The latest push incorporates those changes.